Support for ARK cluster + shared server files

DrPsychick opened the pull request:
* update arkmanager version & url
* symlink `Saved` to individual server (to support shared server files, because arkmanager puts PID files in `Saved`)
* shared server files requires disabling staging directory
* always set `am_ark_AltSaveDirectoryName` and create directory
* new variables
  * LOG_RCONCHAT, ARKCLUSTER, ARKSERVER_SHARED
* switch README to LF

@thmhoag I'm happy to make some adjustments if this PR does not work for you

DrPsychick committed:
feat: support /arkclusters volume
fix: flag LOG_RCONCHAT

DrPsychick committed:
feat: support shared server files (symlink saved + logs)
feat: add example flags/opts
feat: update arkmanager URL+version

DrPsychick committed:
fix: require + create $am_ark_AltSaveDirectoryName

DrPsychick committed:
fix: link only when not exists
info: SHARED requires disabling staging

DrPsychick committed:
fix: check for `ShooterGame` directory only

DrPsychick committed:
fix: keep `Saved` separate per server

DrPsychick committed:
fix: no " in am_* variables
fix: mount .steam/steamapps only

DrPsychick committed:
fix: add migration to startup (move `Saved` to ``/ark/saved`)
fix: bRawSockets deprecated

edausq commented on Dockerfile:
```diff
@@ -2,10 +2,15 @@ FROM thmhoag/steamcmd:latest
+    && rm -rf /var/tmp/*
 
-RUN curl -sL "https://raw.githubusercontent.com/FezVrasta/ark-server-tools/v1.6.54/netinstall.sh" | bash -s steam && \
+RUN curl -sL "https://raw.githubusercontent.com/arkmanager/ark-server-tools/v1.6.57/netinstall.sh" | bash -s steam && \
```
```suggestion
RUN curl -sL https://git.io/arkmanager | bash -s steam && \
```
maybe we could follow the documentation to install the last version

DrPsychick replied:
Thanks for reviewing @edausq, makes totally sense for me.

DrPsychick committed:
always use latest arkmanager

Co-authored-by: Edouard Dausque <github-account@edouard.dausque.net>

thmhoag commented on Dockerfile:
```diff
@@ -2,10 +2,15 @@ FROM thmhoag/steamcmd:latest
+    && rm -rf /var/tmp/*
 
-RUN curl -sL "https://raw.githubusercontent.com/FezVrasta/ark-server-tools/v1.6.54/netinstall.sh" | bash -s steam && \
+RUN curl -sL https://git.io/arkmanager | bash -s steam && \
```
Can we make this configurable via a build variable? The main reason this is hard coded is because I have had issues in the past where changes to `ark-server-tools` were actually breaking changes for this image. 

Another idea might be to add an environment variable that, if set, updates `ark-server-tools` on startup (with the caveat that breaking changes are possible).

DrPsychick replied:
Sure, I understand and goal should be to provide always a working image.

Suggestion: let's work together to make it configurable and have several docker images, with version number and "latest". Then everybody can choose either a versioned image or "latest".

Example for Dockerfile arguments:
```
ARG ARKMANAGER_VERSION=latest
[...]
RUN if [ $ARKMANAGER_VERSION = "latest" ]; then curl -sL https://git.io/arkmanager; else curl -sL ...; fi | bash -s steam && \
[...]
```

Same goes for the base steamcmd image. We can build for multiple Ubuntu versions, like I did in some of my projects, like my (soon to be obsolete) ARK image: https://github.com/DrPsychick/docker-linuxgsm-ark/blob/928590d95e9be79e4175d7b903611244620e72a5/.travis.yml#L10

DrPsychick replied:
So, I got it to work... BUT it seems the arkmanager installer always uses the latest version, so the `AMG_VERSION` variable won't have any effect. Or am I missing something?

```
Step 11/29 : RUN curl -sL https://raw.githubusercontent.com/arkmanager/ark-server-tools/$AMG_VERSION/netinstall.sh | bash -s steam
 ---> Running in 7ff30764ae9d
Getting latest release...
Latest release is v1.6.58
```

thmhoag commented on Dockerfile:
```diff
@@ -23,15 +28,37 @@ RUN echo "%sudo   ALL=(ALL) NOPASSWD:ALL" > /etc/sudoers && \
+#   am_arkflag_crossplay= \
+#   am_arkflag_NoTransferFromFiltering= \
+#   am_arkflag_servergamelog= \
+#   am_arkflag_ForceAllowCaveFlyers= \
```
Could we remove the extra commented out variables here and just make sure they are documented instead?

DrPsychick replied:
👍  will do.

jcirizar commented:
Very interested in this PR, is this going to be merged anytime soon?

thmhoag commented on run.sh:
```diff
@@ -131,9 +155,10 @@ function stop {
+# log from RCON to stdout
+if [[ $LOG_RCONCHAT = true ]]; then
+  bash -c ./log.sh &
+fi
```
I'm liking the changes here, this has been a "todo" for quite some time. 👍

thmhoag commented:
Hey @DrPsychick, thanks for the PR!

Looking at the changes, it looks like the main goal is to basically have one server installation across multiple containers on the same machine?

Also, it's been a while since I've really dug into the semantics of symlinking a folder, but IIRC it doesn't play nice if the container is running on a Windows machine with the file path being linked to a mount. I haven't explicitly declared support for Windows hosts in the readme but I've always tried to support the possibility.

jcirizar commented:
@thmhoag quick question if you can give me a hint, do you think with this PR is possible to run a docker swarm with multiple vps and use one of them as a volume to share data and cluster config so not everything is in the same server?

DrPsychick commented:
> Looking at the changes, it looks like the main goal is to basically have one server installation across multiple containers on the same machine?

Yes and no. The idea is to support sharing the server files, because they take up most of the space and one would likely want to run all servers on the same/latest version. 
With docker, yes all containers must then run on the same host. In a kubernetes environment, the volume can be shared across multiple nodes/hosts.

> Also, it's been a while since I've really dug into the semantics of symlinking a folder, but IIRC it doesn't play nice if the container is running on a Windows machine with the file path being linked to a mount. I haven't explicitly declared support for Windows hosts in the readme but I've always tried to support the possibility.

Ok, I missed that point. To me it honestly does not make sense to run a linux docker on a windows host to run a natively windows supported server 😁  I'd rather use a linux server for such cases - or run ark natively on windows.

DrPsychick commented:
> @thmhoag quick question if you can give me a hint, do you think with this PR is possible to run a docker swarm with multiple vps and use one of them as a volume to share data and cluster config so not everything is in the same server?

I haven't used docker swarm yet. If the shared volume can be connected all hosts it should work. Currently I'm working on running an ARK cluster on kubernetes, but I'm running all containers on the same host and use shared local disk (called "ephemeral" on AWS) storage. With decent network the impact of using a network storage should not have a massive impact.
https://github.com/DrPsychick/ark-cluster-chart

DrPsychick committed:
feat: LOG_RCONCHAT now is the number of seconds between dumping the chat log
feat: using `ARKSERVER_SHARED` will unset `am_arkStagingDir`
fix: document ENV vars and volumes in readme

DrPsychick committed:
feat: support specific versions of arkmanager via `AMG_VERSION`

DrPsychick committed:
feat: support specific versions of arkmanager via `AMG_BUILD` ``AMG_VERSION`

DrPsychick committed:
feat: support specific versions of arkmanager via `AMG_BUILD` ``AMG_VERSION`

DrPsychick committed:
feat: support specific versions of arkmanager via `AMG_BUILD` ``AMG_VERSION`

DrPsychick committed:
No more symlinks (#1)

everything is just mounted
LIST_MOUNTS to check

DrPsychick commented:
New proposal - which I hope also works with Windows: Nothing is symlinked, everything is just mounted separately.

1. default behaviour remains as it is and nothing has changed.
2. if you want to use ARKSERVER_SHARED, you **must** also mount your `Saved` directory (and possibly the `clusters` directory)

I tested it with a script with different scenarios.

edausq commented on Dockerfile:
```diff
@@ -1,12 +1,37 @@
+ARG STEAMCMD_VERSION=latest
+ARG AMG_BUILD=latest
+ARG AMG_VERSION=v1.6.57
+FROM drpsychick/steamcmd:$STEAMCMD_VERSION AS base
```
what do you think of using steamcmd/steamcmd?

DrPsychick replied:
needs some more adjustments as it does not create the `steam` user

DrPsychick replied:
it's based on using `root` user. I'd rather use the existing image for now.
https://github.com/steamcmd/docker/blob/master/dockerfiles/ubuntu-18/Dockerfile

edausq replied:
right, good point, I hadn't caught that

edausq commented on README.md:
```diff
@@ -1,98 +1,115 @@
+| am_ark_RCONPort | `32330` | RCON port |
+| am_arkwarnminutes | `15` | Number of minutes to wait/warn players before updating/restarting |
+| am_arkflag_crossplay | `false` | Allow crossyplay with Players on Epic |
+| ARKCLUSTER | `false` | Symlinks `ShooterGame/Saved/clusters` to the `/arkclusters` volume |
```
not a symlink anymore, right?

DrPsychick replied:
Hmm, seem to have missed one of the last commits... fixed

edausq commented on Dockerfile:
```diff
@@ -23,15 +48,28 @@ RUN echo "%sudo   ALL=(ALL) NOPASSWD:ALL" > /etc/sudoers && \
 VOLUME /ark
+# optionally shared volumes between servers in a cluster
+VOLUME /arkserver
+# mount /arkserver/ShooterGame/Saved seperate for each server
```
could you document the in the readme how to run properly 2 containers part of a cluster?

DrPsychick replied:
I added an example that works out-of-the-box, I hope you can verify that.

DrPsychick replied:
PS: until this is merged, you need to use `IMAGE=drpsychick/arkcluster`

DrPsychick committed:
No more symlinks

mount each directory separately, no symlinks

DrPsychick committed:
fix: use steamcmd/steamcmd
feat: add cluster example to readme

DrPsychick committed:
fix: use drpsychick/steamcmd to test

DrPsychick committed:
fix: use am_arkNoPortDecrement

DrPsychick committed:
fix: using a unique clusterid

DrPsychick committed:
fix: image name for after merge

DrPsychick committed:
feat: description how to test switching servers

edausq commented on run.sh:
```diff
@@ -11,44 +12,72 @@ echo "# Ark Server - " `date`
+
+if [ "$ARKCLUSTER" = "true" ]; then
+  # directory is created when something is mounted to 'clusters'
+  [ -d "$ARKSERVER_SHARED/ShooterGame/Saved" ] && chown steam:steam $ARKSERVER_SHARED/ShooterGame/Saved
```
```suggestion
  [ -d "$ARKSERVER/ShooterGame/Saved" ] && chown steam:steam $ARKSERVER/ShooterGame/Saved
```
Should it be ARKSERVER instead of ARKSERVER_SHARED, for people not using ARKSERVER_SHARED?
Same comment for the few following lines

DrPsychick replied:
So far, `ARKCLUSTER` required using `ARKSERVER_SHARED` - have you tried it without?

If you're change makes it work, then I'll apply the suggestion, @edausq

DrPsychick replied:
did you have a chance to test it @edausq?

edausq replied:
I have tested without ARKCLUSTER nor ARKSERVER_SHARED, and I could make my cluster work without issue, only with a volume.

DrPsychick replied:
Ok, I'll apply your changes and also use a different variable in the following line.

DrPsychick committed:
Update run.sh

Co-authored-by: Edouard Dausque <github-account@edouard.dausque.net>

DrPsychick committed:
fix: use ARKSERVER variable

DrPsychick commented:
@thmhoag all should be clarified and also backwards compatible. Feel free to merge or let me know if you need any other change.

DrPsychick commented:
@thmhoag want to give it a try? After all, if merged and it breaks stuff, it can easily be rolled back...

DrPsychick commented:
Hey @thmhoag , do you want to give this another look?
